src/Processors/QueryPlan/Optimizations/calculateHashTableCacheKeys.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+
+#include <Core/Joins.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/SetSerialization.h>
+#include <Interpreters/TableJoin.h>
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/JoinStepLogical.h>
+#include <Processors/QueryPlan/ReadFromRemote.h>
+#include <Processors/QueryPlan/Serialization.h>
+#include <Processors/QueryPlan/SourceStepWithFilter.h>
+#include <Storages/IStorage.h>
+#include <Common/SipHash.h>
+
+using namespace DB;
+
+namespace
+{
+
+UInt64 calculateHashFromStep(const ReadFromParallelRemoteReplicasStep & source)
+{
+    SipHash hash;
+    hash.update(source.getSerializationName());
+    hash.update(source.getStorageID().getFullTableName());
+    return hash.get64();
+}
+
+UInt64 calculateHashFromStep(const SourceStepWithFilter & read)
+{
+    WriteBufferFromOwnString wbuf;
+    SerializedSetsRegistry registry;
+    IQueryPlanStep::Serialization ctx{.out = wbuf, .registry = registry};
+
+    writeStringBinary(read.getSerializationName(), wbuf);
+    if (const auto & snapshot = read.getStorageSnapshot())
+        writeStringBinary(snapshot->storage.getStorageID().getFullTableName(), wbuf);
+    if (const auto & dag = read.getPrewhereInfo())
+        dag->prewhere_actions.serialize(ctx.out, ctx.registry);
+
+    SipHash hash;
+    hash.update(wbuf.str());
+    return hash.get64();
+}
+
+UInt64 calculateHashFromStep(const ITransformingStep & transform)
+{
+    if (!transform.getTransformTraits().preserves_number_of_rows)
+    {
+        WriteBufferFromOwnString wbuf;
+        SerializedSetsRegistry registry;
+        IQueryPlanStep::Serialization ctx{.out = wbuf, .registry = registry};
+
+        writeStringBinary(transform.getSerializationName(), wbuf);
+        try
+        {
+            transform.serialize(ctx);
+        }
+        catch (const Exception & e)
+        {
+            // Some steps currently missing serialization support. Let's just skip them.
+            if (e.code() != ErrorCodes::NOT_IMPLEMENTED)
```

Comment:
Hmm, that's not nice to rethrow/catch exceptions in normal query execution. Can we avoid it, please?